azkaban-common/src/main/java/azkaban/project/JdbcProjectImpl.java

```diff
@@ -802,6 +804,89 @@ public void uploadProjectProperties(final Project project,
     }
   }
 
+  @Override
+  public synchronized FlowRecommendation createFlowRecommendation(final int projectId, final String flowId)
```

Comment:
And since createPod in K8sImpl also could have call createFlowRecommendation for existing project (if I understand it correctly), better to add duplicate check here, i,e. fetch by project_id, flow_id, if exists, update instead of insert.
If you plan something else, please ignore this one.

Reply:
It will throw duplicate SQLException here so duplicate check is not necessary.
But ideally no duplication will occur b/c I will lock and read in-memory project object whenever operating flow recommendations.

Reply:
I am under the impression that only duplicate on Primary key would throw that exception? Correct me if I am wrong.
I am not seeing the logic now but I trust you will handle this well.

Reply:
I'm capturing SQLException (base sql related exception). It should be thrown and caught for most bad sql issues.